### ISSUE TITLE
pycodestyle config not being read, seeing runtime errors since 0.12.0

### ISSUE THREAD

**Issue by Terr** [palantir/python-language-server#210]
Since upgrading to 0.12.1 (from 0.11.1) I noticed that I was getting D100 and D103 docstring warnings in my code, even though I ignored them via `$HOME/.config/pycodestyle`.  I tried adding a `setup.cfg` to the project directory but I still got the warnings.

When running `pyls` (as `pyls --tcp --port 9001 -v`) in TCP mode I saw the following output after loading a Python file. Could the issue I'm seeing be related to the pydocstyle error? 

```
[...]
2017-12-25 11:09:59,221 UTC - INFO - pyls.config.config - Loaded pyls plugin pycodestyle from <module 'pyls.plugins.pycodestyle_lint' from '/home/terr/.local/lib/python3.5/site-packages/pyls/plugins/pycodestyle_lint.py'>
[...]
2017-12-25 11:04:15,188 UTC - INFO - pyls.config.config - Loaded pyls plugin mccabe from <module 'pyls.plugins.mccabe_lint' from '/home/terr/.local/lib/python3.5/site-packages/pyls/plugins/mccabe_lint.py'>
2017-12-25 11:04:15,188 UTC - INFO - pyls.config.config - Loaded pyls plugin jedi_signature_help from <module 'pyls.plugins.signature' from '/home/terr/.local/lib/python3.5/site-packages/pyls/plugins/signature.py'>
Usage: pydocstyle [options] [<file|dir>...]
pyls: error: no such option: --tcp
```

I'm assuming the "no such option: --tcp" error doesn't occur in IO mode. I've got the most recent version of `pydocstyle` installed, 2.1.1. 

I also tried 0.12.0, where I got the same output and results (warnings about D100/D103). Version 0.11.1 still works fine, both in IO and TCP mode.

**Comment by gatesn** [comment]
I think what's happening is pydocstyle is reading sys.argv. Will look into a work-around there for you. 

Could you try this branch regarding the warnings you thought were disabled by your setup.cfg? ngates/pycodestyle-ignores

**Comment by gerardroche** [comment]
I have the same issue in 0.12.2.


This the flake8 file:

```
[flake8]
max-line-length = 120

ignore = D100,D101,D102,D103,D104,D105,D107
```

Here are the warnings (there should be no D100 or D103 because they're ignored in the flake8 file):
 
```
      10:1   	pycodestyle 	warning   	E302 expected 2 blank lines, found 1
       12:1   	pycodestyle 	warning   	E302 expected 2 blank lines, found 0
       15:1   	pycodestyle 	warning   	E302 expected 2 blank lines, found 1
       19:1   	pycodestyle 	warning   	E302 expected 2 blank lines, found 1
        1:1   	pydocstyle  	warning   	D100: Missing docstring in public module
       10:1   	pydocstyle  	warning   	D103: Missing docstring in public function
       12:1   	pydocstyle  	warning   	D103: Missing docstring in public function
       15:1   	pydocstyle  	warning   	D103: Missing docstring in public function
      102:1   	pydocstyle  	warning   	D103: Missing docstring in public function
      108:1   	pydocstyle  	warning   	D103: Missing docstring in public function
```

Here's some mc logging:

```

2018-01-11 19:01:18,837 UTC - DEBUG - pyls.server - Got message: b'{"jsonrpc": "2.0", "method": "initialized", "params": {}}':
2018-01-11 19:01:18,837 UTC - DEBUG - pyls.server - Got message: b'{"jsonrpc": "2.0", "method": "workspace/didChangeConfiguration", "params": {"settings": {"pyls": {"configurationSources": ["flake8", "pycodestyle"]}}}}':
2018-01-11 19:01:18,838 UTC - INFO - pyls.config.config - Updated settings to {'configurationSources': ['flake8', 'pycodestyle']}:
2018-01-11 19:01:18,838 UTC - INFO - pyls.config.config - Disabled plugins: []:
2018-01-11 19:01:18,845 UTC - DEBUG - pyls.config.config -   pyls_document_did_open [hook]:
2018-01-11 19:01:18,845 UTC - DEBUG - pyls.config.config -   finish pyls_document_did_open --> [] [hook]:
2018-01-11 19:01:19,347 UTC - DEBUG - pyls.config.config -   pyls_lint [hook]:
2018-01-11 19:01:19,347 UTC - DEBUG - pyls.config.config - Got user config from PyCodeStyleConfig: {}:
2018-01-11 19:01:19,348 UTC - DEBUG - pyls.config.config - Got user config from Flake8Config: {}:
2018-01-11 19:01:19,348 UTC - DEBUG - pyls.config.config - With user configuration: {}:
2018-01-11 19:01:19,349 UTC - DEBUG - pyls.config.config - With plugin configuration: {'plugins': {'pydocstyle': {'enabled': False}}}:
2018-01-11 19:01:19,349 UTC - DEBUG - pyls.config.config - With lsp configuration: {'configurationSources': ['flake8', 'pycodestyle'], 'plugins': {'pydocstyle': {'enabled': False}}}:
2018-01-11 19:01:19,349 UTC - DEBUG - pyls.config.source - Using cached configuration for ():
2018-01-11 19:01:19,350 UTC - DEBUG - pyls.config.config - Got project config from PyCodeStyleConfig: {}:
2018-01-11 19:01:19,351 UTC - DEBUG - pyls.config.config - Got project config from Flake8Config: {'plugins': {'pycodestyle': {'ignore': ['D100', 'D101', 'D102', 'D103', 'D104', 'D105', 'D107'], 'maxLineLength': 120}}}:
2018-01-11 19:01:19,351 UTC - DEBUG - pyls.config.config - With project configuration: {'configurationSources': ['flake8', 'pycodestyle'], 'plugins': {'pycodestyle': {'ignore': ['D100', 'D101', 'D102', 'D103', 'D104', 'D105', 'D107'], 'maxLineLength': 120}, 'pydocstyle': {'enabled': False}}}:
2018-01-11 19:01:19,351 UTC - DEBUG - pyls.plugins.pycodestyle_lint - Got pycodestyle settings: {'ignore': ['D100', 'D101', 'D102', 'D103', 'D104', 'D105', 'D107'], 'maxLineLength': 120}:
2018-01-11 19:01:19,422 UTC - INFO - pyls.plugins.pydocstyle_lint - Got error: /home/code/sandbox/file.py:1 at module level::
D100: Missing docstring in public module:
2018-01-11 19:01:19,422 UTC - INFO - pyls.plugins.pydocstyle_lint - Got error: /home/code/sandbox/file.py:10 in public function `function_name`::
D103: Missing docstring in public function:
2018-01-11 19:01:19,423 UTC - INFO - pyls.plugins.pydocstyle_lint - Got error: /home/code/sandbox/file.py:12 in public function `function_name`::
D103: Missing docstring in public function:
2018-01-11 19:01:19,423 UTC - INFO - pyls.plugins.pydocstyle_lint - Got error: /home/code/sandbox/file.py:15 in public function `function_name`::
D103: Missing docstring in public function:
2018-01-11 19:01:19,424 UTC - INFO - pyls.plugins.pydocstyle_lint - Got error: /home/code/sandbox/file.py:102 in public function `function_name`::
D103: Missing docstring in public function:
2018-01-11 19:01:19,424 UTC - INFO - pyls.plugins.pydocstyle_lint - Got error: /home/code/sandbox/file.py:108 in public function `function_name`::
D103: Missing docstring in public function:
2018-01-11 19:01:19,425 UTC - DEBUG - pyls.config.source - Using cached configuration for ():
2018-01-11 19:01:19,425 UTC - DEBUG - pyls.config.config - Got user config from PyCodeStyleConfig: {}:
2018-01-11 19:01:19,425 UTC - DEBUG - pyls.config.source - Using cached configuration for ():
2018-01-11 19:01:19,425 UTC - DEBUG - pyls.config.config - Got user config from Flake8Config: {}:
2018-01-11 19:01:19,425 UTC - DEBUG - pyls.config.config - With user configuration: {}:
2018-01-11 19:01:19,425 UTC - DEBUG - pyls.config.config - With plugin configuration: {'plugins': {'pydocstyle': {'enabled': False}}}:
2018-01-11 19:01:19,425 UTC - DEBUG - pyls.config.config - With lsp configuration: {'configurationSources': ['flake8', 'pycodestyle'], 'plugins': {'pydocstyle': {'enabled': False}}}:
2018-01-11 19:01:19,425 UTC - DEBUG - pyls.config.source - Using cached configuration for ():
2018-01-11 19:01:19,426 UTC - DEBUG - pyls.config.config - Got project config from PyCodeStyleConfig: {}:
2018-01-11 19:01:19,426 UTC - DEBUG - pyls.config.source - Using cached configuration for ('/home/code/sandbox/.flake8',):
2018-01-11 19:01:19,426 UTC - DEBUG - pyls.config.config - Got project config from Flake8Config: {'plugins': {'pycodestyle': {'ignore': ['D100', 'D101', 'D102', 'D103', 'D104', 'D105', 'D107'], 'maxLineLength': 120}}}:
2018-01-11 19:01:19,426 UTC - DEBUG - pyls.config.config - With project configuration: {'configurationSources': ['flake8', 'pycodestyle'], 'plugins': {'pycodestyle': {'ignore': ['D100', 'D101', 'D102', 'D103', 'D104', 'D105', 'D107'], 'maxLineLength': 120}, 'pydocstyle': {'enabled': False}}}:
2018-01-11 19:01:19,426 UTC - DEBUG - pyls.plugins.mccabe_lint - Running mccabe lint with threshold: 15:
2018-01-11 19:01:19,428 UTC - DEBUG - pyls.config.config -   finish pyls_lint --> [[{'source': 'pycodestyle', 'code': 'E302', 'severity': 2, 'range': {'start': {'line': 9, 'character': 0}, 'end': {'line': 9, 'character': 26}}, 'message': 'E302 expected 2 blank lines, found 1'}, {'source': 'pycodestyle', 'code': 'E302', 'severity': 2, 'range': {'start': {'line': 11, 'character': 0}, 'end': {'line': 11, 'character': 25}}, 'message': 'E302 expected 2 blank lines, found 0'}, {'source': 'pycodestyle', 'code': 'E302', 'severity': 2, 'range': {'start': {'line': 14, 'character': 0}, 'end': {'line': 14, 'character': 18}}, 'message': 'E302 expected 2 blank lines, found 1'}, {'source': 'pycodestyle', 'code': 'E302', 'severity': 2, 'range': {'start': {'line': 18, 'character': 0}, 'end': {'line': 18, 'character': 17}}, 'message': 'E302 expected 2 blank lines, found 1'}], [], [{'source': 'pydocstyle', 'message': 'D100: Missing docstring in public module', 'severity': 2, 'range': {'start': {'line': 0, 'character': 0}, 'end': {'line': 0, 'character': 49}}, 'code': 'D100'}, {'source': 'pydocstyle', 'message': 'D103: Missing docstring in public function', 'severity': 2, 'range': {'start': {'line': 9, 'character': 0}, 'end': {'line': 9, 'character': 49}}, 'code': 'D103'}, {'source': 'pydocstyle', 'message': 'D103: Missing docstring in public function', 'severity': 2, 'range': {'start': {'line': 11, 'character': 0}, 'end': {'line': 11, 'character': 49}}, 'code': 'D103'}, {'source': 'pydocstyle', 'message': 'D103: Missing docstring in public function', 'severity': 2, 'range': {'start': {'line': 14, 'character': 0}, 'end': {'line': 14, 'character': 49}}, 'code': 'D103'}, {'source': 'pydocstyle', 'message': 'D103: Missing docstring in public function', 'severity': 2, 'range': {'start': {'line': 101, 'character': 0}, 'end': {'line': 101, 'character': 49}}, 'code': 'D103'}, {'source': 'pydocstyle', 'message': 'D103: Missing docstring in public function', 'severity': 2, 'range': {'start': {'line': 107, 'character': 0}, 'end': {'line': 107, 'character': 49}}, 'code': 'D103'}], []] [hook]:
```

Trackback to https://github.com/tomv564/LSP/issues/239

**Comment by Terr** [comment]
@gatesn I've tried your branch and the `ignore` setting is being used now (tested with W292, no newline at end of file). Thanks for that!

I was still getting the 'DXXX' notices despite being ignored, like in my original message, so I decided to dive a bit deeper. I just discovered that `pycodestyle` isn't responsible for checking for 'DXXX' violations at all. So it was just a coincidence that there was an actual issue with warnings/errors not being ignored by `pycodestyle` since 0.12.0 :sweat_smile:

So to summarize: with your branch the `ignore` setting is working again. `pydocstyle` is still having issues in TCP mode but works correctly in stdio mode.
